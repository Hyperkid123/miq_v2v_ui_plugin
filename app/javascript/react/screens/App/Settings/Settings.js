import React from 'react';
import PropTypes from 'prop-types';
import { reduxForm, Field } from 'redux-form';
import { Breadcrumb, Form, Grid, Button } from 'patternfly-react';
import Toolbar from '../../../config/Toolbar';
import NumberInput from '../common/forms/NumberInput';

const normalizeStringToInt = str => str && parseInt(str.replace(/\D/g, ''), 10) || 0;

class Settings extends React.Component {
  componentDidMount() {
    const { fetchSettingsAction } = this.props;
    fetchSettingsAction('TODO: url here'); // TODO
  }

  onApplyClick = () => {
    console.log('TODO: apply settings changes');
  };

  render() {
    const { settingsForm } = this.props;

    const toolbarContent = (
      <Toolbar>
        <Breadcrumb.Item href="/dashboard/maintab?tab=compute">{__('Compute')}</Breadcrumb.Item>
        <Breadcrumb.Item href="/migration">{__('Migration')}</Breadcrumb.Item>
        <Breadcrumb.Item active>{__('Migration Settings')}</Breadcrumb.Item>
      </Toolbar>
    );

    const settingsContent = (
      <div className="migration-settings">
        <h2>{__('Concurrent Migrations')}</h2>
        <Form horizontal>
          <Form.FormGroup>
            <Grid.Col componentClass={Form.ControlLabel} sm={5}>
              {__('Maximum Concurrent migrations per conversion host')}
            </Grid.Col>
            <Grid.Col sm={7}>
              <div style={{ width: 100 }}>
                <Field
                  id="maxMigrationsPerHost"
                  name="maxMigrationsPerHost"
                  component={NumberInput}
                  normalize={normalizeStringToInt}
                />
              </div>
            </Grid.Col>
          </Form.FormGroup>
          <Form.FormGroup>
            <Grid.Col sm={12}>
              <Button bsStyle="primary" onClick={this.onApplyClick}>
                {__('Apply')}
              </Button>
            </Grid.Col>
          </Form.FormGroup>
        </Form>
      </div>
    );

    return (
      <React.Fragment>
        {toolbarContent}
        {settingsContent}
      </React.Fragment>
    );
  }
}

Settings.propTypes = {
  fetchSettingsAction: PropTypes.func,
  settingsForm: PropTypes.object
};

export default reduxForm({
  form: 'settings'
})(Settings);
