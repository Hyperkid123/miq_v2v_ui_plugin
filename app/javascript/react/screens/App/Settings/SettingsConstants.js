export const V2V_FETCH_SERVERS = 'V2V_FETCH_SERVERS';
export const V2V_FETCH_SETTINGS = 'V2V_FETCH_SETTINGS';
export const V2V_PATCH_SETTINGS = 'V2V_PATCH_SETTINGS';
export const FETCH_V2V_CONVERSION_HOSTS = 'FETCH_V2V_CONVERSION_HOSTS';
