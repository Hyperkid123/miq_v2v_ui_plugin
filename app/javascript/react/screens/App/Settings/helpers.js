export const getFormValuesFromApiSettings = payload => ({
  max_concurrent_tasks_per_host: payload.transformation.limits.max_concurrent_tasks_per_host
});

export const getApiSettingsFromFormValues = values => ({
  transformation: {
    limits: {
      max_concurrent_tasks_per_host: values.max_concurrent_tasks_per_host
    }
  }
});

// Example task name: "Configuring a conversion_host: operation=enable resource=(type: ManageIQ::Providers::Openstack::CloudManager::Vm id:42000000000113)"
const conversionHostTaskNameRegex = /operation=(\w+)\s+resource=\(type:\s+([\w:]+)\s+id:(\d+)\)/;

export const parseConversionHostTasksMetadata = tasks =>
  tasks &&
  tasks.map(task => {
    const result = conversionHostTaskNameRegex.exec(task.name);
    return {
      ...task,
      nameMeta: {
        operation: result[1],
        resourceType: result[2],
        resourceId: result[3]
      }
    };
  });
