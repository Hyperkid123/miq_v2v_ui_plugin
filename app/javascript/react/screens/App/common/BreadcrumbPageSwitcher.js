import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { OverlayTrigger, Popover, Button, Icon, ListGroup, ListGroupItem } from 'patternfly-react';

const BreadcrumbPageSwitcher = ({ items, activeHref }) => {
  const popover = (
    <Popover id="breadcrumb-switcher-popover">
      <ListGroup style={{ marginBottom: 0, borderTop: 0 }}>
        {items.map(({ name, href }) => (
          <ListGroupItem className={classNames('no-border', { active: href === activeHref })} href={href} key={href}>
            {name}
          </ListGroupItem>
        ))}
      </ListGroup>
    </Popover>
  );

  return (
    <OverlayTrigger rootClose trigger="click" placement="bottom" overlay={popover}>
      <Button
        style={{
          marginTop: '-2px',
          marginBottom: '-4px',
          marginLeft: '15px',
          height: '22px',
          width: '32px',
          lineHeight: '14px'
        }}
      >
        <Icon type="fa" name="exchange" />
      </Button>
    </OverlayTrigger>
  );
};

BreadcrumbPageSwitcher.propTypes = {
  activeHref: PropTypes.string.isRequired,
  items: PropTypes.arrayOf(PropTypes.shape({ name: PropTypes.string, href: PropTypes.string })).isRequired
};

export default BreadcrumbPageSwitcher;
