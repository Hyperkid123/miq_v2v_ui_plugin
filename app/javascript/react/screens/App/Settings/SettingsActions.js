import URI from 'urijs';
// import API from '../../../../common/API';

import { V2V_FETCH_SETTINGS } from './SettingsConstants';

const _getSettingsActionCreator = url => dispatch =>
  dispatch({
    type: V2V_FETCH_SETTINGS,
    // payload: API.get(url)
    // TODO replace this dummy payload with actual settings from API
    payload: new Promise(resolve => {
      setTimeout(
        () =>
          resolve({
            maxMigrationsPerHost: 10
          }),
        2000
      );
    })
  });

export const fetchSettingsAction = url => {
  const uri = new URI(url);
  return _getSettingsActionCreator(uri.toString());
};
