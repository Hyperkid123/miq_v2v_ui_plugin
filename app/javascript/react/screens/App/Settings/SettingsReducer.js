import Immutable from 'seamless-immutable';

import { V2V_FETCH_SETTINGS, V2V_POST_SETTINGS } from './SettingsConstants';

export const initialState = Immutable({
  isFetchingSettings: false,
  fetchingSettingsRejected: false,
  errorFetchingSettings: null,
  savedSettings: {},
  isSavingSettings: false,
  savingSettingsRejected: false,
  errorSavingSettings: null
});

export default (state = initialState, action) => {
  switch (action.type) {
    case `${V2V_FETCH_SETTINGS}_PENDING`:
      return state
        .set('isFetchingSettings', true)
        .set('fetchingSettingsRejected', false)
        .set('errorFetchingSettings', null);
    case `${V2V_FETCH_SETTINGS}_REJECTED`:
      return state
        .set('isFetchingSettings', false)
        .set('fetchingSettingsRejected', true)
        .set('errorFetchingSettings', action.payload);
    case `${V2V_FETCH_SETTINGS}_FULFILLED`:
      return state
        .set('isFetchingSettings', false)
        .set('fetchingSettingsRejected', false)
        .set('errorFetchingSettings', null)
        .set('savedSettings', action.payload);

    case `${V2V_POST_SETTINGS}_PENDING`:
      return state
        .set('isSavingSettings', true)
        .set('savingSettingsRejected', false)
        .set('errorSavingSettings', null);
    case `${V2V_POST_SETTINGS}_REJECTED`:
      return state
        .set('isSavingSettings', false)
        .set('savingSettingsRejected', true)
        .set('errorSavingSettings', action.payload);
    case `${V2V_POST_SETTINGS}_FULFILLED`:
      return state
        .set('isSavingSettings', false)
        .set('savingSettingsRejected', false)
        .set('errorSavingSettings', null)
        .set('savedSettings', action.payload);

    default:
      return state;
  }
};
