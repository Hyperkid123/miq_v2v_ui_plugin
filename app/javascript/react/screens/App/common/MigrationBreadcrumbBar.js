import React from 'react';
import PropTypes from 'prop-types';
import { Breadcrumb } from 'patternfly-react';
import Toolbar from '../../../config/Toolbar';
import BreadcrumbPageSwitcher from './BreadcrumbPageSwitcher';

const items = [
  { name: __('Migration Plans'), href: '#/plans' },
  { name: __('Infrastructure Mappings'), href: '#/mappings' },
  { name: __('Migration Settings'), href: '#/settings' }
];

const MigrationBreadcrumbBar = ({ activeHref }) => {
  const activeItem = items.find(item => item.href === activeHref);
  return (
    <Toolbar>
      <Breadcrumb.Item active>{__('Migration')}</Breadcrumb.Item>
      <Breadcrumb.Item active>
        <strong>{activeItem.name}</strong>
      </Breadcrumb.Item>
      <BreadcrumbPageSwitcher items={items} activeHref={activeHref} />
    </Toolbar>
  );
};

MigrationBreadcrumbBar.propTypes = {
  activeHref: PropTypes.string.isRequired
};

export default MigrationBreadcrumbBar;
