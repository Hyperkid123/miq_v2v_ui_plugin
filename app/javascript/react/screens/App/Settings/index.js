import { connect } from 'react-redux';
import Settings from './Settings';

import * as SettingsActions from './SettingsActions';
import * as NotificationActions from '../common/NotificationList/NotificationListActions';

import reducer from './SettingsReducer';

export const reducers = { settings: reducer, form: {} };

const mapStateToProps = ({ settings }, ownProps) => ({
  ...settings,
  ...ownProps.data
});

const mergeProps = (stateProps, dispatchProps, ownProps) => Object.assign(stateProps, ownProps.data, dispatchProps);

export default connect(
  mapStateToProps,
  Object.assign(SettingsActions, NotificationActions),
  mergeProps
)(Settings);
