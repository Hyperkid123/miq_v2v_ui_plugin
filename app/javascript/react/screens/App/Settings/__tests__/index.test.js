import React from 'react';
import thunk from 'redux-thunk';
import promiseMiddleware from 'redux-promise-middleware';
import { createStore, applyMiddleware, combineReducers } from 'redux';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { reducer as formReducer } from 'redux-form';

import Settings from '../Settings';
import SettingsContainer, { reducers } from '../index';
import { initialState } from '../SettingsReducer';
import { servers, settings } from '../settings.fixures';

import { mockRequest } from '../../../../../common/mockRequests';

describe('Settings integration test', () => {
  global.$ = jest.fn(() => ({
    TouchSpin: jest.fn(),
    on: jest.fn(),
    off: jest.fn()
  }));
  mockRequest({
    url: '/api/servers',
    response: { data: servers }
  });
  mockRequest({
    url: '/api/settings',
    response: { data: settings }
  });

  const middlewares = [thunk, promiseMiddleware()];
  const generateStore = () =>
    createStore(
      combineReducers({
        ...reducers,
        form: formReducer
      }),
      {
        settings: initialState
      },
      applyMiddleware(...middlewares)
    );

  const mountComponent = store =>
    mount(
      <Provider store={store}>
        <SettingsContainer />
      </Provider>
    );

  it('should mount the Settings page with mapStateToProps reduced', () => {
    const store = generateStore();
    const wrapper = mountComponent(store);
    const component = wrapper.find(Settings);
    expect(component.props()).toMatchSnapshot();
  });
});
