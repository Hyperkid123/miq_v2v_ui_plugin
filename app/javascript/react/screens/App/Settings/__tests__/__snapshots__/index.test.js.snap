// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings integration test should mount the Settings page with mapStateToProps reduced 1`] = `
Object {
  "addNotificationAction": [Function],
  "enableReinitialize": true,
  "errorFetchingServers": null,
  "errorFetchingSettings": null,
  "errorSavingSettings": null,
  "fetchServersAction": [Function],
  "fetchSettingsAction": [Function],
  "fetchingServersRejected": false,
  "fetchingSettingsRejected": false,
  "initialValues": Object {},
  "isFetchingServers": true,
  "isFetchingSettings": true,
  "isSavingSettings": false,
  "patchSettingsAction": [Function],
  "removeNotificationAction": [Function],
  "savedSettings": Object {},
  "savingSettingsRejected": false,
  "servers": Array [],
  "settingsForm": Object {
    "initial": Object {},
    "values": Object {},
  },
}
`;
